containers.go

```diff
@@ -141,16 +141,21 @@ type rwContainerStore interface {
 }
 
 type containerStore struct {
-	lockfile   *lockfile.LockFile
-	dir        string
-	jsonPath   [numContainerLocationIndex]string
+	// The following fields are only set when constructing containerStore, and must never be modified afterwards.
+	// They are safe to access without any other locking.
+	lockfile *lockfile.LockFile // Synchronizes readers vs. writers of the _filesystem data_, both cross-process and in-process.
+	dir      string
+	jsonPath [numContainerLocationIndex]string
+
+	inProcessLock sync.RWMutex // Can _only_ be obtained with lockfile held.
+	// The following fields can only be read/written with read/write ownership of inProcessLock, respectively.
+	// Almost all users should use startReading() or startWriting().
```

Comment:
Absolutely non-blocking and just an idea.  The `Image` struct in libimage has some fields that may be cached.  I think you (@mtrmac) suggested to create an embedded struct to highlight the caching nature of these fields (see https://github.com/containers/common/blob/main/libimage/image.go#L38). Cached fields are always accessed via `(*Image).cached.field` which IMO makes the code much more expressive and comprehensible.

Would it help to add the fields that require locking in a similar embedded struct? `(*containerStore).requiresLock.containers`?

Reply:
Ooh, that did not occur to me at all, and it’s definitely worth doing.

Reply:
I have prototyped this, and it does seem valuable — but it’s a lot of churn.

E.g. on `containers.go`, the current PR is +80-14 non-whitespace non-comment changes (including some changes with comments), and moving the fields around would be about +51-45 . Also, in addition to C/I/L stores, the top-level `store` would benefit from the same updates, but `store.go` is mostly out of scope of this PR.

Overall, I’d prefer to do it in a separate PR a bit later: it would help maintenance but it should not have immediately user-visible impact, which makes it harder for me to prioritize right now — and the work so far churn has already been disrupting Nalin’s work, so coordinating a bit better seems warranted.

I have added the suggestion to #1389; it might well be worth tracking as a higher-priority issue separate from the somewhat desperate #1389.

Reply:
Yes, let's do that in a separate PR. Thank you!